libdirections-hybrid/src/test/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouterTest.kt

```diff
@@ -0,0 +1,187 @@
+        verify(exactly = 1) { offboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 1) { routerCallback.onResponse(any()) }
+    }
+
+    @Test
+    fun whenOffboardRouterFailsOnboardRouterIsCalledAndOffboardUsedAgain() = rule.runBlockingTest {
+        enableNetworkConnection()
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        internalCallback.captured.onFailure(Throwable())
+        internalCallback.captured.onResponse(emptyList())
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        internalCallback.captured.onResponse(emptyList())
+
+        verify(exactly = 2) { offboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 1) { onboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 2) { routerCallback.onResponse(any()) }
+    }
+
+    @Test
+    fun whenOnboardRouterFailsOffboardRouterIsCalledAndOnboardUsedAgain() = rule.runBlockingTest {
+        disableNetworkConnection()
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        internalCallback.captured.onFailure(Throwable())
+        internalCallback.captured.onResponse(emptyList())
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        internalCallback.captured.onResponse(emptyList())
+
+        verify(exactly = 2) { onboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 1) { offboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 2) { routerCallback.onResponse(any()) }
+    }
+
+    @Test
+    fun whenConnectionAppearedRoutersSwitched() = rule.runBlockingTest {
+        disableNetworkConnection()
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        enableNetworkConnection()
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        verify(exactly = 1) { onboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 1) { offboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+    }
+
+    @Test
+    fun whenConnectionDisappearedRoutersSwitched() = rule.runBlockingTest {
+        enableNetworkConnection()
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        disableNetworkConnection()
+
+        hybridRouter.getRoute(routerOptions, routerCallback)
+
+        verify(exactly = 1) { offboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+        verify(exactly = 1) { onboardRouter.getRoute(routerOptions, capture(internalCallback)) }
+    }
+
+    private fun enableNetworkConnection() = networkConnected(true)
+
+    private fun disableNetworkConnection() = networkConnected(false)
+
+    private fun networkConnected(networkConnected: Boolean) {
+        every { networkInfo.isConnectedOrConnecting } returns networkConnected
+
+        receiver.captured.onReceive(context, Intent())
+
+        Thread.sleep(1000)
```

Comment:
Could this be workarounded? It's not a good practice to add `Thread.sleep` into tests 😬

Reply:
channel receives updates from receiver asynchronously, I didn't find I way to make tests stable without `sleep`

Reply:
Are there any tools in the coroutines testing lib?

You could possibly also use the `CountDownLatch` or anything like that.

Reply:
the problem is that I don't touch coroutines in that test, I just mock a callback. 
Also I don't touch `NetworkStatusService` where channel is stored

Reply:
Does `Channel#offer` synchronously deliver data to observers? If yes, you could inject `NetworkStatusService` via internal constructor while the public one creates it itself. This way you'd have a way of controlling the channels state. You could also take the same tactic with the `NetworkStatusService`'s callback, inject it via an internal constructor and notify it whenever you need.

Reply:
`Channel#offer` is synchronous, but we listen channel inside a coroutine. And it is suspended. So, If we need to add any delay (with `Thread.sleep` or `delay()`) to allow our channel listener to receive `network state` before other code runs.
You are right, I can provide channel listener in constructor and call it directly when I need. 
I used `Thread.sleep` logic because in that case I don't touch inner components (channel, listener), I just call `BroadcastReceivers` callback with a new state. So I know nothing about channels. But I have to use thread lock. 
What do you think? @LukasPaczos @RingerJK @Guardiola31337